opacus/optimizer.py

```diff
@@ -0,0 +1,118 @@
+    if std > 0:
+        return torch.normal(
+            mean=0,
+            std=std,
+            size=reference.shape,
+            device=reference.device,
+        )
+    return torch.zeros(reference.shape, device=reference.device)
+
+
+class DPOptimizer(Optimizer):
+
+    def __init__(self,
+                 optimizer: Optimizer,
+                 *,
+                 noise_multiplier: float,
+                 max_grad_norm: float,
+                 expected_batch_size: Optional[int],
+                 loss_reduction: str = "mean",
+    ):
+        if loss_reduction not in ("mean", "sum"):
+            raise ValueError(f"Unexpected value for loss_reduction: {loss_reduction}")
+
+        if loss_reduction == "mean" and expected_batch_size is None:
+            raise ValueError("You must provide expected batch size of the loss reduction is mean")
+
+        self.optimizer = optimizer
+        self.noise_multiplier = noise_multiplier
+        self.max_grad_norm = max_grad_norm
+        self.loss_reduction = loss_reduction
+        self.expected_batch_size = expected_batch_size
+        self.step_hook = None
+
+        self.accumulated_iterations = 0
+
+    @property
+    def params(self) -> List[nn.Parameter]:
+        ret = []
+        for param_group in self.optimizer.param_groups:
+            ret += [p for p in param_group['params'] if p.requires_grad]
+        return ret
+
+    @property
+    def grad_samples(self) -> List[torch.Tensor]:
+        ret = []
+        for p in self.params:
+            if not hasattr(p, "grad_sample"):
+                raise ValueError("Per sample gradient not found. Are you using GradSampleModule?")
+
+            ret.append(p.grad_sample)
+        return ret
+
+    def attach_step_hook(self, fn: Callable[[DPOptimizer], None]):
+        self.step_hook = fn
+
+    def clip_and_accumulate(self):
+        per_param_norms = [x.view(len(x), -1).norm(2, dim=-1) for x in self.grad_samples]
+        per_sample_norms = torch.stack(per_param_norms, dim=1).norm(2, dim=1)
+        per_sample_clip_factor = (self.max_grad_norm / (per_sample_norms + 1e-6)).clamp(max=1.0)
+
+        for p in self.params:
+            grad = torch.einsum("i,i...", per_sample_clip_factor, p.grad_sample)
+
+            if hasattr(p, "summed_grad"):
+                p.summed_grad += grad
+            else:
+                p.summed_grad = grad
+
+    def add_noise(self):
+        for p in self.params:
+            noise = _generate_noise(self.noise_multiplier * self.max_grad_norm, p.summed_grad)
+            p.grad = p.summed_grad + noise
+
+    def scale_grad(self):
+        if self.loss_reduction == "mean":
+            for p in self.params:
+                p.grad /= (self.expected_batch_size * self.accumulated_iterations)
+
+    # TODO: see GradSampleModule.zero_grad()
+    # TODO: actually, not calling zero_grad() after step() does break privacy accounting - add warning?
+    def zero_grad(self, set_to_none: bool = False):
```

Comment:
wait a second. I started questioning the premise of my #TODO. Does it break accounting?

As long as we reuse the gradient after we've added noise, it should be fine: it can be considered a post-processing, which is fine under DP guarantees. 

It looks to me as if, as long as we don't accumulate raw gradients between steps, we're good.
Is there any flaw in this logic?

Reply:
> If you take two steps without zero_grad, in the second one you would have twice as much data points but only add noise that covers one batch worth of data

True, thanks.

>  I think that the closest to Pytorch would be to accumulate in p.grad

The problem with that is PyTorch will add regular non-clipped gradients to the same field. We had to accumulate in `summed_grad` because we always need to override p.grad as the last step.
What if we do
```
    def add_noise(self):
        for p in self.params:
            noise = _generate_noise(<...>)
            p.summed_grad += noise
            p.grad = p.summed_grad
```

?

Reply:
There's additional complication of scaling - we need to divide the gradients by batch size.
I think the following could work:

we accumulate raw, non-noised, non-scaled, but clipped gradients in `p.clipped_grad`
on every `step()` we do 
```
p.summed_grad += (p.clipped_grad + noise) / (self.expected_batch_size * self.accumulated_iterations)
p.grad = p.summed_grad
del p.clipped_grad
```